File: clients/go/bashclient/observe.go
package bashclient

import (
	"context"
	"sync"
)

// ObserveChannel handles channel 3: command observation events.
type ObserveChannel struct {
	send sendFunc
	recv recvFunc

	mu          sync.Mutex
	onPreCmd    func(PreCommandEvent)
	onPostCmd   func(PostCommandEvent)
}

func newObserveChannel(send sendFunc, recv recvFunc) *ObserveChannel {
	return &ObserveChannel{send: send, recv: recv}
}

// Subscribe subscribes to command observation events at the given level.
func (c *ObserveChannel) Subscribe(ctx context.Context, level int) error {
	if err := c.send(MakeMsg(ChanObserve, "subscribe", map[string]any{"level": level})); err != nil {
		return err
	}
	resp, err := c.recv(ctx, ChanObserve)
	if err != nil {
		return err
	}
	if getString(resp, "type") == "error" {
		return newServerError(getString(resp, "message"), ChanObserve)
	}
	return nil
}

// Unsubscribe stops observation.
func (c *ObserveChannel) Unsubscribe(ctx context.Context) error {
	if err := c.send(MakeMsg(ChanObserve, "unsubscribe")); err != nil {
		return err
	}
	resp, err := c.recv(ctx, ChanObserve)
	if err != nil {
		return err
	}
	if getString(resp, "type") == "error" {
		return newServerError(getString(resp, "message"), ChanObserve)
	}
	return nil
}

// OnPreCommand registers a callback for pre_command events. Pass nil to unregister.
func (c *ObserveChannel) OnPreCommand(fn func(PreCommandEvent)) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.onPreCmd = fn
}

// OnPostCommand registers a callback for post_command events. Pass nil to unregister.
func (c *ObserveChannel) OnPostCommand(fn func(PostCommandEvent)) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.onPostCmd = fn
}

// dispatch routes a server-push message to registered callbacks.
func (c *ObserveChannel) dispatch(msg Message) {
	t := getString(msg, "type")
	data := getMap(msg, "data")
	if data == nil {
		data = msg
	}

	c.mu.Lock()
	preCb := c.onPreCmd
	postCb := c.onPostCmd
	c.mu.Unlock()

	switch t {
	case "pre_command":
		if preCb != nil {
			preCb(PreCommandEvent{
				Seq:        getInt(msg, "seq"),
				Timestamp:  getInt64(msg, "timestamp"),
				Command:    getString(data, "command"),
				Cwd:        getString(data, "cwd"),
				LineNumber: getInt(data, "line_number"),
				IsSubshell: getBool(data, "is_subshell"),
				IsAsync:    getBool(data, "is_async"),
			})
		}
	case "post_command":
		if postCb != nil {
			postCb(PostCommandEvent{
				Seq:          getInt(msg, "seq"),
				Timestamp:    getInt64(msg, "timestamp"),
				Command:      getString(data, "command"),
				ExitStatus:   getInt(data, "exit_status"),
				SignalNumber: getInt(data, "signal_number"),
				DurationMs:   getInt(data, "duration_ms"),
			})
		}
	}
}

File: bash-server/clients/typescript/src/channels/observe.ts
import { ServerError } from "../errors";
import { makeMsg } from "../protocol";
import {
  CHAN_OBSERVE, Message, PreCommandEvent, PostCommandEvent,
} from "../types";

type SendFn = (msg: Message) => Promise<void>;
type RecvFn = (ch: number) => Promise<Message>;
type Callback = (...args: any[]) => void;

export class ObserveChannel {
  private callbacks: Map<string, Callback[]> = new Map();

  constructor(private send: SendFn, private recv: RecvFn) {}

  async subscribe(level: number = 1): Promise<void> {
    await this.send(makeMsg(CHAN_OBSERVE, "subscribe", { level }));
    const resp = await this.recv(CHAN_OBSERVE);
    if (resp.type === "error") {
      throw new ServerError(String(resp.message ?? "subscribe error"), CHAN_OBSERVE);
    }
  }

  async unsubscribe(): Promise<void> {
    await this.send(makeMsg(CHAN_OBSERVE, "unsubscribe"));
    const resp = await this.recv(CHAN_OBSERVE);
    if (resp.type === "error") {
      throw new ServerError(String(resp.message ?? "unsubscribe error"), CHAN_OBSERVE);
    }
  }

  on(event: string, callback: Callback): void {
    const list = this.callbacks.get(event) ?? [];
    list.push(callback);
    this.callbacks.set(event, list);
  }

  off(event: string, callback?: Callback): void {
    if (!callback) {
      this.callbacks.delete(event);
    } else {
      const list = this.callbacks.get(event) ?? [];
      this.callbacks.set(event, list.filter((cb) => cb !== callback));
    }
  }

  /** Dispatch a server-push message. */
  dispatch(msg: Message): void {
    const t = String(msg.type ?? "");
    const data = (msg.data ?? {}) as Record<string, unknown>;

    let event: PreCommandEvent | PostCommandEvent | Message;
    if (t === "pre_command") {
      event = {
        seq: Number(msg.seq ?? 0),
        timestamp: Number(msg.timestamp ?? 0),
        command: String(data.command ?? ""),
        cwd: String(data.cwd ?? ""),
        line_number: Number(data.line_number ?? 0),
        is_subshell: Boolean(data.is_subshell),
        is_async: Boolean(data.is_async),
      };
    } else if (t === "post_command") {
      event = {
        seq: Number(msg.seq ?? 0),
        timestamp: Number(msg.timestamp ?? 0),
        command: String(data.command ?? ""),
        exit_status: Number(data.exit_status ?? 0),
        signal_number: Number(data.signal_number ?? 0),
        duration_ms: Number(data.duration_ms ?? 0),
      };
    } else {
      event = msg;
    }

    for (const cb of this.callbacks.get(t) ?? []) {
      cb(event);
    }
  }
}
